src/annotate.rs

```diff
@@ -37,65 +37,92 @@ pub enum MemberId<'a> {
 pub trait Annotate {
     fn format(&self, variant: Option<&str>, field: &MemberId) -> Option<Format>;
     fn comment(&self, variant: Option<&str>, field: &MemberId) -> Option<String>;
+    fn as_annotate(&self) -> Option<&dyn Annotate>;
+    fn thunk_serialize(&self, serializer: &mut AnnotatedSerializer) -> Result<Document, Error>;
 }
 
-type IdFn = fn() -> usize;
-type CastFn = unsafe fn(*const ()) -> &'static dyn Annotate;
+/// The default implementation of Annotate returns no comments or annotations and
+/// cannot return the trait object.
+impl<T: ?Sized + serde::Serialize> Annotate for T {
+    default fn format(&self, _variant: Option<&str>, _field: &MemberId) -> Option<Format> {
+        None
+    }
+    default fn comment(&self, _variant: Option<&str>, _field: &MemberId) -> Option<String> {
+        None
+    }
+    default fn as_annotate(&self) -> Option<&dyn Annotate> {
+        None
+    }
+    default fn thunk_serialize(
+        &self,
+        serializer: &mut AnnotatedSerializer,
+    ) -> Result<Document, Error> {
+        self.serialize(serializer)
+    }
+}
 
-pub struct AnnotateType {
-    pub id: IdFn,
-    pub cast: CastFn,
+// We use a private trait to identify whether the Serializer passed to
+// serde::Serialize for dyn Annotate is AnnotatedSerializer.
+trait IsAnnotatedSerializer {
+    fn is_annotated_serializer(&self) -> bool;
 }
-inventory::collect!(AnnotateType);
 
-impl AnnotateType {
-    pub fn type_id<T>() -> usize
-    where
-        T: ?Sized,
-    {
-        // Just like https://github.com/rust-lang/rust/issues/41875#issuecomment-317292888
-        // We monomorphize on T and then cast the function pointer address of
-        // the monomorphized `AnnotateType::type_id` function to an
-        // integer identifier.
-        Self::type_id::<T> as usize
+impl<T: serde::Serializer> IsAnnotatedSerializer for T {
+    default fn is_annotated_serializer(&self) -> bool {
+        false
     }
-
-    pub unsafe fn cast<T>(ptr: *const ()) -> &'static dyn Annotate
-    where
-        T: 'static + Annotate,
-    {
-        // Cast a generic pointer back to a reference to T and return a
-        // dyn reference to the Annotate trait.
-        &*(ptr as *const T)
+}
+impl<'a> IsAnnotatedSerializer for AnnotatedSerializer<'a> {
+    fn is_annotated_serializer(&self) -> bool {
+        true
+    }
+}
```

Comment:
I don't think this first implementation is correct. I believe it will result in trying to transmute a `AnnotatedSerializer` directly into a `&mut AnnotatedSerializer`.

I tried commenting it out and `cargo +nightly-2022-07-03 test` still passes, so it seems to be unnecessary?

Reply:
I meant to leave a comment asking a question about that (ie: why do I have to impl for `&mut AnnotatedSerializer<'a>`?)

But it's just as well - that's the only type `Serialize for dyn Annotate` can accept anyway.

Reply:
Yeah the references are kinda tricky here.